src/org/jitsi/impl/neomedia/transform/fec/FlexFec03Receiver.java

```diff
@@ -0,0 +1,372 @@
+        = Logger.getLogger(FlexFec03Receiver.class);
+
+    /**
+     * FEC-related statistics
+     */
+    private Statistics statistics;
+
+    /**
+     * Helper class to reconstruct missing packets
+     */
+    private Reconstructor reconstructor;
+
+    public FlexFec03Receiver(long mediaSsrc, byte fecPayloadType)
+    {
+        super(mediaSsrc, fecPayloadType);
+        this.statistics = new Statistics();
+        this.reconstructor = new Reconstructor(mediaPackets);
+    }
+
+    @Override
+    public synchronized RawPacket[] doReverseTransform(RawPacket[] pkts)
+    {
+        Set<Integer> flexFecPacketsToRemove = new HashSet<>();
+        // Try to recover any missing media packets
+        for (Map.Entry<Integer, RawPacket> entry : fecPackets.entrySet())
+        {
+            FlexFec03Packet flexFecPacket = FlexFec03Packet.create(entry.getValue());
+            if (flexFecPacket == null)
+            {
+                continue;
+            }
+            reconstructor.setFecPacket(flexFecPacket);
+            if (reconstructor.complete())
+            {
+                flexFecPacketsToRemove.add(flexFecPacket.getSequenceNumber());
+                continue;
+            }
+            if (reconstructor.canRecover())
+            {
+                logger.debug("Attemping recovery of missing sequence " +
+                    "number " + reconstructor.missingSequenceNumber);
+                flexFecPacketsToRemove.add(flexFecPacket.getSequenceNumber());
+                RawPacket recovered = reconstructor.recover();
+                if (recovered != null)
+                {
+                    logger.info("Recovered packet " +
+                        recovered.getSequenceNumber());
+                    statistics.numRecoveredPackets++;
+                    saveMedia(recovered);
+                    pkts = insert(recovered, pkts);
+                }
+                else
+                {
+                    logger.error("Recovery of packet " +
+                        reconstructor.missingSequenceNumber + " failed even" +
+                        " though it should have been possible");
+                    statistics.failedRecoveries++;
+                }
+            }
+        }
+
+        for (Integer flexFecSeqNum : flexFecPacketsToRemove)
+        {
+            fecPackets.remove(flexFecSeqNum);
+        }
+        return pkts;
+    }
+
+    /**
+     * Inserts packet into an empty slot in pkts, or allocates a new
+     * array and inserts packet into it.  Returns either the original
+     * array (with packet insert) or a new array containing the original contents
+     * of pkts and with packet inserted
+     * @param packet the packet to be inserted
+     * @param pkts the array in which to insert packet
+     * @return the original pkts array with packet inserted, or, a new array
+     * containing all elements in pkts as well as packet
+     */
+    private RawPacket[] insert(RawPacket packet, RawPacket[] pkts)
```

Comment:
I think that'd be a nice addition to ArrayUtils

Reply:
it looks like `ArrayUtils#add` does something similar, but it doesn't cover the case where there is an existing empty slot that can be used.  not sure if i'd want to mess with `ArrayUtils#add`, should i add this method alongside `ArrayUtils#add`? what do you think?

Reply:
I was thinking of a new method in ArrayUtils, which would use ArrayUtils#add if there is no empty space.

Reply:
good idea, done.